folding-schemes/src/folding/hypernova/circuits.rs

```diff
@@ -1051,43 +1149,132 @@ mod tests {
+            if i == 0 {
+                // hash the initial (dummy) CycleFold instance, which is used as the 2nd public
+                // input in the AugmentedFCircuit
+                let cf_u_i1_x = cf_U_i.hash_cyclefold(&poseidon_config).unwrap();
+
+                augmented_f_circuit =
+                    AugmentedFCircuit::<Projective, Projective2, GVar2, CubicFCircuit<Fr>> {
+                        _c2: PhantomData,
+                        _gc2: PhantomData,
+                        poseidon_config: poseidon_config.clone(),
+                        ccs: ccs.clone(),
+                        i: Some(iFr),
+                        i_usize: Some(i),
+                        z_0: Some(z_0.clone()),
+                        z_i: Some(z_i.clone()),
+                        external_inputs: Some(vec![]),
+                        u_i_C: Some(u_i.C),
+                        U_i: Some(U_i.clone()),
+                        U_i1_C: Some(U_i1.C),
+                        F: F_circuit,
+                        x: Some(u_i1_x),
+                        nimfs_proof: Some(nimfs_proof),
+
+                        // cyclefold values
+                        cf_u_i_cmW: None,
+                        cf_U_i: None,
+                        cf_x: Some(cf_u_i1_x),
+                        cf_cmT: None,
+                    };
+            } else {
+                let rho_Fq = Fq::from_bigint(BigInteger::from_bits_le(&rho_bits)).unwrap();
+                // CycleFold part:
+                // get the vector used as public inputs 'x' in the CycleFold circuit
+                // cyclefold circuit for cmW
+                let cf_u_i_x = [
+                    vec![rho_Fq],
+                    get_cm_coordinates(&U_i.C),
+                    get_cm_coordinates(&u_i.C),
+                    get_cm_coordinates(&U_i1.C),
+                ]
+                .concat();
+
+                let cf_circuit = CycleFoldCircuit::<Projective, GVar> {
+                    _gc: PhantomData,
+                    r_bits: Some(rho_bits.clone()),
+                    p1: Some(U_i.clone().C),
+                    p2: Some(u_i.clone().C),
+                    x: Some(cf_u_i_x.clone()),
                 };
 
+                let (_cf_w_i, cf_u_i, cf_W_i1, cf_U_i1, cf_cmT, _) = fold_cyclefold_circuit::<
+                    Projective,
+                    GVar,
+                    Projective2,
+                    GVar2,
+                    CubicFCircuit<Fr>,
+                    Pedersen<Projective>,
+                    Pedersen<Projective2>,
+                >(
+                    &poseidon_config,
+                    cf_r1cs.clone(),
+                    cf_pedersen_params.clone(),
+                    cf_W_i.clone(), // CycleFold running instance witness
+                    cf_U_i.clone(), // CycleFold running instance
+                    cf_u_i_x,       // CycleFold incoming instance
+                    cf_circuit,
+                )
+                .unwrap();
```

Comment:
Also here, should we go for `expect` or err handling?

Reply:
but this is a test

Reply:
ohh nevermind...